### PR TITLE
num_rows=None -> num_rows=?

### LINKED ISSUE #29779: [Datasets] repr shows num_rows=None
### What happened + What you expected to happen

When a Dataset does not know the number of rows, the Dataset repr will sometimes show `num_rows=None`. The intent is to say that "the number of rows is unknown", but it may be confused with "there are no rows".

### Versions / Dependencies

ray latest, python 3.7, mac intel

### Reproduction script

(thanks to @clarkzinzow)
```
import ray
ds = ray.data.read_csv(["example://iris.csv", "example://iris.csv"])
print(ds)
# --> Dataset(num_blocks=2, num_rows=None, schema={sepal.length: double, sepal.width: double, petal.length: double, petal.width: double, variety: string})
```

### Issue Severity

Low: It annoys or frustrates me.

### PR DESCRIPTION
**Pull request by xcharleslin**
Signed-off-by: Xiayue Charles Lin <xcl@anyscale.com>

## Why are these changes needed?

Dataset repr shows `Datasets(num_rows=None)` when the number of rows is unknown. This may be misconstrued as no rows. `Datasets(num_rows=?)` is more accurate.

## Related issue number

Closes #29779 .

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by ericl — APPROVED** [review]
Update unit tests as well?

File: python/ray/data/dataset.py [hunk 1]
```diff
@@ -4000,8 +4000,9 @@ def __repr__(self) -> str:
             schema_str = ", ".join(schema_str)
             schema_str = "{" + schema_str + "}"
         count = self._meta_count()
+        count_str = str(count) if count is not None else "?"
         return "Dataset(num_blocks={}, num_rows={}, schema={})".format(
-            self._plan.initial_num_blocks(), count, schema_str
+            self._plan.initial_num_blocks(), count_str, schema_str
```
Comment (clarkzinzow) [inline python/ray/data/dataset.py:4005]: Nit: `str(count)` will already happen implicitly when formatting the string, so we can just replace `count` with `"?"` if it's `None`.
```python
count = self._meta_count()
if count is None:
    count = "?"
return "Dataset(num_blocks={}, num_rows={}, schema={})".format(
    self._plan.initial_num_blocks(), count, schema_str
)

**Review by clarkzinzow — COMMENTED** [review]
LGTM overall, +1 on updating tests

**Review by c21 — APPROVED** [review]
LGTM

**Pull request merged**